### PR TITLE
Added alternative login providers for desura  **Requires desura web update to work**

### PR DESCRIPTION
**Pull request by lodle**
Added staging option for login/api services to allow testing against non production server
Added the ability to debug api calls to server with xdebug
Disabled cert check for staging url
Moved all url generation to webcore
Fixed password box being empty after login failure #637
Fixed asserts from wxWidgets (wrong thread) after login ext failure

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,7 +3,7 @@
 CMakeCache.txt
 CMakeFiles/
 Makefile
-build/
+build*/
```
Comment (Jookia) [inline .gitignore:6]: Will break the build_out folder.

File: build_out/data/language/english.xml [hunk 2]
```diff
@@ -768,5 +768,7 @@
 
 		<str name="#BH_CRASH">Browser has died. (Click to reload)</str>
 		<str name="#BH_NO_LOAD">Failed to load browser library. Please check console for error message.</str>
+		
+		<str name="#ALTLOGIN_TITLE">Alternative Login</str>
```
Comment (Jookia) [inline build_out/data/language/english.xml:772]: Make sure you update https://github.com/desura/Desurium/wiki/Translation-TODO .

File: build_out/data/themes/default/theme.xml [hunk 3]
```diff
@@ -278,7 +278,15 @@
...
-		<image name="chat_profile_nonactive">chat_profile_nonactive.png</image>			
+		<image name="chat_profile_nonactive">chat_profile_nonactive.png</image>
+		
+		<image name="test_good">test_good.png</image>
+		<image name="test_bad">test_bad.png</image>
+		
+		<image name="login_twitter">login_twitter.png</image>
+		<image name="login_steam">login_steam.png</image>
+		<image name="login_facebook">login_facebook.png</image>
+		<image name="login_google">login_google.png</image>
```
Comment (Jookia) [inline build_out/data/themes/default/theme.xml:289]: Maybe we need a theme TODO as well.

File: src/shared/uicore/CMakeLists.txt [hunk 4]
```diff
@@ -120,7 +120,8 @@ file(GLOB Sources       code/AboutForm.cpp
                         code/UploadProgPage.cpp
                         code/UploadPrompt.cpp
                         code/UsernameBox.cpp
-                        code/*.h)
+                        code/*.h
+			code/AltLoginDialog.cpp)
```
Comment (Jookia) [inline src/shared/uicore/CMakeLists.txt:124]: Could you put this before the header and align it?

File: src/shared/uicore/code/AltLoginDialog.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+AltLoginDialog::AltLoginDialog(wxWindow* parent, const char* szProvider, const char* szApiUrl)
+	: gcDialog(parent, wxID_ANY, wxT("#ALTLOGIN_TITLE"), wxDefaultPosition, wxSize(500, 665), wxDEFAULT_FRAME_STYLE|wxTAB_TRAVERSAL) 
+{
+	SetTitle(gcString("{0}: {1}", Managers::GetString(L"#ALTLOGIN_TITLE"), szProvider));
+
+	GetWebCore();
+
+	gcString strUrl;
+	
+	if (gcString(szApiUrl).size() > 0)
```
Comment (Jookia) [inline src/shared/uicore/code/AltLoginDialog.cpp:32]: Is this for internal backwards compatibility?

File: .gitignore (hunk 1 above)
Comment (lodle) [inline .gitignore:6]: I know, not the biggest issue as windows copies into that folder atm. Need to rethink this though

File: src/shared/uicore/code/ItemUpdateForm.cpp [hunk 6]
```diff
@@ -108,7 +108,7 @@ void UpdateInfoForm::setInfo(DesuraId id, bool launch)
 
 	SetLabel(gcWString(L"{0} {1}", Managers::GetString(L"#UI_TITLE"), item->getName()));
 
-	m_ieBrowser->loadUrl(gcWString(L"http://www.desura.com/{0}/{1}/changelog", id.getTypeString(), item->getShortName()));
+	m_ieBrowser->loadUrl(gcWString(L"{0}/{1}/{2}/changelog", GetWebCore()->getUrl(WebCore::Root), id.getTypeString(), item->getShortName()));
```
Comment (Jookia) [inline src/shared/uicore/code/ItemUpdateForm.cpp:111]: Maybe we should start documenting the URL layout?

File: src/shared/uicore/code/LoginForm.cpp [hunk 7]
```diff
@@ -877,6 +1002,47 @@ void LoginForm::newAccountLogin(const char* username, const char* cookie)
...
+	const char* szProvider = NULL;
+
+	if (event.GetId() == m_butTwitter->GetId())
+		szProvider = "twitter";
+	else if (event.GetId() == m_butSteam->GetId())
+		szProvider = "steam";
+	else if (event.GetId() == m_butFacebook->GetId())
+		szProvider = "facebook";
+	else if (event.GetId() == m_butGoogle->GetId())
+		szProvider = "google";
```
Comment (Jookia) [inline src/shared/uicore/code/LoginForm.cpp:1028]: It seems we're doing this in a lot of areas. Would it be worth refactoring in to a table of some kind?

File: src/shared/uicore/CMakeLists.txt (hunk 4 above)
Comment (lodle) [inline src/shared/uicore/CMakeLists.txt:124]: Damn tabs and spaces

File: src/shared/uicore/code/gcWCUtil.cpp [hunk 8]
```diff
@@ -235,13 +235,23 @@ void ShutdownWebControl()
...
-	CEF_DeleteCookie("http://www.desura.com", "masterchief");
+	gcString urlRoot;
+
+	{
+		Thread::AutoLock a(g_RootUrlMutex);
+		urlRoot = g_strRootUrl;
+	}
+
+	CEF_DeleteCookie(urlRoot.c_str(), "freeman");
+	CEF_DeleteCookie(urlRoot.c_str(), "masterchief");
```
Comment (Jookia) [inline src/shared/uicore/code/gcWCUtil.cpp:254]: Legacy crud?

Comment (lodle) [inline src/shared/uicore/code/gcWCUtil.cpp:254]: Huh? this is needed so the user is logged in on desura.com

Comment (Jookia) [inline src/shared/uicore/code/gcWCUtil.cpp:254]: Ah, so desura uses those names as cookies? I thought it was some testing stuff.

File: src/shared/uicore/code/LoginForm.cpp [hunk 9]
```diff
@@ -730,6 +838,9 @@ void LoginForm::onLogin()
 		gc_lastusername.setValue((const char*)dastr.c_str());
 	}
 
+	if (m_comboProvider)
+		gc_login_stage_last.setValue(m_comboProvider->GetStringSelection());
```
Comment (Jookia) [inline src/shared/uicore/code/LoginForm.cpp]: This causes a compile error.

Comment (Jookia) [inline src/shared/uicore/code/LoginForm.cpp]: To fix, add `.ToUTF8()`

File: src/shared/uicore/code/LoginForm.cpp [hunk 10]
```diff
@@ -730,6 +838,9 @@ void LoginForm::onLogin()
 		gc_lastusername.setValue((const char*)dastr.c_str());
```
Comment (Jookia) [inline src/shared/uicore/code/LoginForm.cpp]: This should be `dastr.ToUTF8()`

File: src/shared/uicore/code/LoginForm.cpp (hunk 9 above)
Comment (lodle) [inline src/shared/uicore/code/LoginForm.cpp]: Should be fixed

File: src/shared/uicore/code/AltLoginDialog.cpp (hunk 5 above)
Comment (lodle) [inline src/shared/uicore/code/AltLoginDialog.cpp:32]: Which part?

Comment (Jookia) [inline src/shared/uicore/code/AltLoginDialog.cpp:32]: The two cases here, instead of one.

**Comment by lodle** [comment]
Redoing

**Pull request closed without merging**